docs/developer/api_guidelines.md

```diff
@@ -0,0 +1,438 @@
+* `operations` - internal API providing non-functional features like statistics, debug and testing never exposed by the API gateway
+
+## Public APIs
+Public APIs are always provided by the API gateway and mapped to the K8s Service as follows:
+
+API Gateway: `http://cluster.keptn.sh/api/<component>/<version>/<api resources>`
+
+K8S Service: `<component>/public/<version>/<api resources>`
+
+## Internal APIs
+Internal APIs are not exposed by the API gateway are accessed by other services in Keptn like follows:
+
+K8S Service:  `<component>/internal/<version>/<api resources>`
+
+## Operations APIs
+Operations APIs are not exposed by the API gateway and **should** cover non-functional features
+like statistics, debug or testing APIs etc...
+
+K8S Service: `<component>/operations/<version>/<api resources>` 
+
+---
+
+# Authentication
+Each request **must** have a *Authorization* header containing a base64 encoded bearer string:
+
+## Example
+
+```
+GET /controlplane/v1/sequences
+Authorization: Bearer dGhlc2VjcmV0dG9rZW5mb3JrZXB0bmFwaQ==
+```
+
+Internal communication does not require a token.
+
+---
+
+# Custom Methods
+
+Sometimes, functionality cannot be expressed using the default methods on classical REST resources.
+For this use cases *custom methods* **should** be used:
+
+```
+https://domain/api/<component>/<version>/<api resources>:<custom-method>
+```
+
+- Custom methods **should** use the `HTTP POST` method.
+- The URL **must** end with a verb separated with a colon.
+
+## Example
+```
+POST cluster.keptn.sh/api/controlplane/v1/sequences/94dbe764-e186-48bf-b90d-a53b21fca847:pause
+POST cluster.keptn.sh/api/tokens/v1/token:rotate
+```
+
+ ---
+# Filtering
+
+API endpoints supporting result filtering **must** accept the filter
+as a query parameter named *filter* containing a *filter expression*:
+
+In general each field **should** be filter-able. Exceptions **must** be documented.
+*Filter expressions* follow the following form: `<fieldname> operator <value>`
+
+## Filters
+
+### Filter based on numbers
+
+**Operators:**
+
+```
+= != < <= > >=
+```
+
+**Format**:
+
+```
+int - decimal and hexadecimal (leading 0x)`
+float - scientific notation (optional exponent e or E) 
+```
+
+**Example:**
+
+```
+failedAttempt >= 3
+average >= 1.0E2
+responseTimeSec > 0.23
+```
+
+## Filter based on strings
+
+**Operators:**
+
+```
+= != contains, starts-with, ends-with
+```
+
+**Format**:
+
+```
+strings wrapped in single quotes
+special characters escaped with \
+```
+
+**Example:**
+
+```
+projectName = 'myproject'
+```
+
+## Filter based on booleans
+
+**Operators:**
+
+```
+= !=
+```
+
+**Format**:
+
+```
+true false
+```
+
+**Example:**
+
+```
+`active != true`
+```
+
+## Filter based on Date / Time
+
+**Operators:**
+
+```
+= != < <= > >=
+```
+
+**Format**:
+
+```
+ISO 8601 string
+```
+
+**Example:**
+
+```
+lastModified >= '2020-01-01T01:00:00Z
+```
+
+---
+# JSON Format
+
```

Comment:
Maybe add a sentence about Accept and Content-Type headers that MUST be `application/json`